ROADMAP.md

```diff
@@ -0,0 +1,24 @@
+# etcd roadmap
+
+This document defines high level goals for project.
+
+## Milestones
+
+* [P0] Etcd releases are qualified by rigorous robustness testing
+* [P0] Etcd can reliably detect data corruption
```

Comment:
How does the above `P1` 3 action items fit into the Milestones section here? 

> etcd apply code should be easy to understand and validate correctness
> etcd can reliably detect data corruption (hash is linearizable)
(This P1 action item is mentioned as P0)

> etcd recovery from data inconsistency procedures are documented and tested

How does the stalled write due to slow disk fit into the Milestone? @ahrtr 

What about lease redesign? https://github.com/etcd-io/etcd/pull/14094

Reply:
> etcd apply code should be easy to understand and validate correctness

I would want to avoid touching random parts of apply code without better testing and clear goal. Removal of v2 API and following cleanup should already improve the situation and takes priority.

 > etcd can reliably detect data corruption (hash is linearizable)
 
 This should be covered by ` Etcd can reliably detect data corruption`, maybe requires rewrite/clarification. I might downgrade it back to P1 as it doesn't help with v3.4 release.
 
 > etcd recovery from data inconsistency procedures are documented and tested.
 
 Documentation is done, we just need to add test. I think we should file an issue as important but goal milestone itself should be tracked as part of improvements to testing.
 
 > How does the stalled write due to slow disk fit into the Milestone? @ahrtr

This is somewhat new effort that is still not well defined. For me it comes under reliability, which is important, but as it relates to hardware failures it's not something etcd tackled yet. However with recently reported https://github.com/etcd-io/etcd/issues/15498 I would want to propose "etcd is resilient to hardware failures" soon.

> What about lease redesign? 

Correctness should be our top priority, however leases have been broken for long time and no-one cared (K8s also doesn't). As so I would treat it second priority to KV API.
At some point we could consider an larger effort "etcd APIs is high quality and has consistent behavior" that would encompass leases.

Reply:
Is the scope for this item clear? IIRC there was a discussion on corruption detection per key/value, then there were some discussions around merkle trees and partitioning the keyspace.

Reply:
> Is the scope for this item clear? IIRC there was a discussion on corruption detection per key/value, then there were some discussions around merkle trees and partitioning the keyspace.

If you mean corruption detection scope, then not. I didn't have time to define it. It's pretty large issue to tackle and are multiple ways to approach it. Main challenge is balancing breaking changes and short term vs long term improvements. I have couple of ideas that I discussed with @ptabor, but didn't have time to write them down as I want to focus on finishing robustness tests first (not too long). 

Happy to make the scope clearer if someone is interested in working on it. Still would like to encourage people to work on robustness tests first, as they also help v3.4 and v3.5 releases.

Reply:
is there any tracking issue for changes to catch corruption per key/value?